Added more feature in "Dark Mode".  i.e. Scheduling Dark Mode

Coderxrohan opened the pull request:
## 📝 Description
I have made changes to automate the dark mode. Means the user can be able to schedule the dark mode as his choice.

## 📸 Screenshots / 📹 Videos
Attached is the screenshot below.

## 🔗 Related Issues
- I wish the scheduling should pop up when the user enables DAKR mode; otherwise, the scheduling option should be hidden. so it does not consume any sugar.

## ✅ Checklist
[✅] I have read and followed the [Contributing Guidelines](https://github.com/XengShi/materialYouNewTab/
blob/main/CONTRIBUTING.md).
![Screenshot 2025-01-25 191436](https://github.com/user-attachments/assets/66c12374-af7a-442c-989a-e04d58495f45)
![Screenshot 2025-01-25 191445](https://github.com/user-attachments/assets/dfbb58c1-8afc-4c99-b41c-11e517fbc51b)
[✅] I have tested my changes by installing them as an extension (not just running via localhost) to ensure it builds, installs, and works as expected.
[✅] I have tested these changes in at least Chrome and Firefox (other browsers if applicable).
[✅] Attached visual evidence of changes (screenshots or videos), if applicable.

Coderxrohan commented:
I made changes and also worked on this further.
1) Scheduling should be enabled when the dark mode setting is enabled. Otherwise, the scheduling should be hidden in the dark mode toggle. So this will consume space in the panel.

Coderxrohan commented:
@prem-k-r  and  @Thunder-Blaze.
Have a look please :)

prem-k-r commented:
Hi Rohan, can you please move all newly added code in script.js to dark-mode.js

prem-k-r commented on index.html:
```diff
@@ -1033,7 +1033,7 @@ <h1>Material You NewTab</h1>
                                         style="transform: scale(1.1); transform-origin: center;">
                                         <path
-                                            d="M12 21q-.425 0-.712-.288T11 20v-4q0-.425.288-.712T12 15t.713.288T13 16v1h7q.425 0 .713.288T21 18t-.288.713T20 19h-7v1q0 .425-.288.713T12 21m-8-2q-.425 0-.712-.288T3 18t.288-.712T4 17h4q.425 0 .713.288T9 18t-.288.713T8 19zm4-4q-.425 0-.712-.288T7 14v-1H4q-.425 0-.712-.288T3 12t.288-.712T4 11h3v-1q0-.425.288-.712T8 9t.713.288T9 10v4q0 .425-.288.713T8 1 ...
+                                            d="M12 21q-.425 0-.712-.288T11 20v-4q0-.425.288-.712T12 15t.713.288T13 16v1h7q.425 0 .713.288T21 18t-.288.713T20 19h-7v1q0 .425-.288.713T16 21m-8-2q-.425 0-.712-.288T3 18t.288-.712T4 17h4q.425 0 .713.288T9 18t-.288.713T8 19zm4-4q-.425 0-.712-.288T7 14v-1H4q-.425 0-.712-.288T3 12t.288-.712T4 11h3v-1q0-.425.288-.712T8 9t.713.288T9 10v4q0 .425-.288.713T8 1 ...
```
Here this change in svg is intentional or mistake?

Coderxrohan replied:
By mistake

Coderxrohan commented:
> Hi Rohan, can you please move all newly added code in script.js to dark-mode.js

I will make a new dark mode file, and move that block of code to that file.

prem-k-r commented:
Not like that.
there will be no addition in script.js

in dark-mode.js
```
document.addEventListener("DOMContentLoaded", function () {
 // Add these new constants with the other element selectors
    const scheduledDarkModeCheckbox = document.getElementById("scheduleDarkModeCheckbox");
    const scheduledDarkModeField = document.getElementById("scheduledDarkModeField");

    // Load saved schedule preferences
    const isDarkModeScheduled = localStorage.getItem("darkModeScheduled") === "true";
    scheduledDarkModeCheckbox.checked = isDarkModeScheduled;

    // Enable/disable schedule checkbox based on dark mode state
    enableDarkModeCheckbox.addEventListener("change", function() {
        const isDarkModeEnabled = this.check
      
      ... ADD REST
      
      // Initial check on page load
    if (isDarkModeScheduled) {
        scheduledDarkModeCheckbox.disabled = !enableDarkModeCheckbox.checked;
        checkAndApplyScheduledDarkMode();
    }
});

// Add these constants at the top with your other constants
const enableDarkModeCheckbox = document.getElementById("enableDarkModeCheckbox");
const scheduledDarkModeCheckbox = document.getElementById("scheduleDarkModeCheckbox");

// Add this HTML after the schedule checkbox in your settings menu
const scheduleSettingsHTML = `
    <div id="scheduleSettings" class="schedule-settings" style="display: none;">
        <div class
        
          ... ADD REST
          
          .time-input::-webkit-calendar-picker-indicator {
            filter: var(--calendar-icon-filter);
        }
    `;
    document.head.appendChild(style);
});
```
also html and css part here can be moved to index.html and style.css

Coderxrohan commented:
I made the new dark-mode.js file and
```html
// Dark Mode Scheduling
class DarkModeScheduler {
    constructor() {
        this.scheduleEnabled = false;
        this.startTime = '20:00'; // Default start time (8 PM)
        this.endTime = '06:00';   // Default end time (6 AM)
        
        this.init();
    }

    init() {
        // Load saved settings
        this.loadSettings();
        
        // Initialize UI elements
        this.scheduleToggle = document.getElementById('darkModeSchedule');
        this.scheduleSettings = document.getElementById('darkModeScheduleSettings');
        this.startTimeInput = document.getElementById('darkModeStart');
        this.endTimeInput = document.getElementById('darkModeEnd');
        this.darkModeCheckbox = document.getElementById('enableDarkModeCheckbox');

        // Set up event listeners
        this.scheduleToggle.addEventListener('change', () => this.toggleSchedule());
        this.startTimeInput.addEventListener('change', () => this.updateSchedule());
        this.endTimeInput.addEventListener('change', () => this.updateSchedule());

        // Initialize UI state
        this.scheduleToggle.checked = this.scheduleEnabled;
        this.scheduleSettings.style.display = this.scheduleEnabled ? 'block' : 'none';
        this.startTimeInput.value = this.startTime;
        this.endTimeInput.value = this.endTime;

        // Start checking schedule
        if (this.scheduleEnabled) {
            this.checkSchedule();
            this.startScheduleCheck();
        }
    }

    loadSettings() {
        const settings = JSON.parse(localStorage.getItem('darkModeSettings') || '{}');
        this.scheduleEnabled = settings.enabled || false;
        this.startTime = settings.startTime || '20:00';
        this.endTime = settings.endTime || '06:00';
    }

    saveSettings() {
        const settings = {
            enabled: this.scheduleEnabled,
            startTime: this.startTime,
            endTime: this.endTime
        };
        localStorage.setItem('darkModeSettings', JSON.stringify(settings));
    }

    toggleSchedule() {
        this.scheduleEnabled = this.scheduleToggle.checked;
        this.scheduleSettings.style.display = this.scheduleEnabled ? 'block' : 'none';
        
        if (this.scheduleEnabled) {
            this.checkSchedule();
            this.startScheduleCheck();
        } else {
            this.stopScheduleCheck();
            // Reset to manual dark mode state
            this.setDarkMode(this.darkModeCheckbox.checked);
        }
        
        this.saveSettings();
    }

    updateSchedule() {
        this.startTime = this.startTimeInput.value;
        this.endTime = this.endTimeInput.value;
        this.checkSchedule();
        this.saveSettings();
    }

    checkSchedule() {
        if (!this.scheduleEnabled) return;

        const now = new Date();
        const currentTime = `${String(now.getHours()).padStart(2, '0')}:${String(now.getMinutes()).padStart(2, '0')}`;
        const shouldBeDark = this.isTimeBetween(currentTime, this.startTime, this.endTime);
        
        this.setDarkMode(shouldBeDark);
    }

    setDarkMode(enabled) {
        // Update the checkbox state
        this.darkModeCheckbox.checked = enabled;
        
        // Trigger the dark mode change
        const event = new Event('change');
        this.darkModeCheckbox.dispatchEvent(event);
        
        // Apply dark mode directly
        if (enabled) {
            document.documentElement.classList.add('dark-theme');
            document.body.setAttribute('data-bg', 'color');
        } else {
            document.documentElement.classList.remove('dark-theme');
            document.body.setAttribute('data-bg', 'color');
        }
    }

    isTimeBetween(current, start, end) {
        const currentMinutes = this.timeToMinutes(current);
        const startMinutes = this.timeToMinutes(start);
        const endMinutes = this.timeToMinutes(end);

        if (startMinutes <= endMinutes) {
            return currentMinutes >= startMinutes && currentMinutes <= endMinutes;
        } else {
            // Handles cases where the time range crosses midnight
            return currentMinutes >= startMinutes || currentMinutes <= endMinutes;
        }
    }

    timeToMinutes(time) {
        const [hours, minutes] = time.split(':').map(Number);
        return hours * 60 + minutes;
    }

    startScheduleCheck() {
        this.scheduleInterval = setInterval(() => this.checkSchedule(), 60000); // Check every minute
    }

    stopScheduleCheck() {
        if (this.scheduleInterval) {
            clearInterval(this.scheduleInterval);
        }
    }
}

// Initialize dark mode scheduler when the document is loaded
document.addEventListener('DOMContentLoaded', () => {
    window.darkModeScheduler = new DarkModeScheduler();
}); 
```   moved that code in this js.
Moved the unwanted part to css.
As you said.
Take a look.

Coderxrohan commented:
See the working 

https://github.com/user-attachments/assets/5def1368-0ff1-47fa-a774-75f88fdad3e3

prem-k-r commented:
Yes, it working, Great work! 
But there are lots of unnecessary codes

Thunder-Blaze commented:
> Yes, it working, Great work! 
> But there are lots of unnecessary codes

Yeah I thought same but I didn't tested it so idk if they are for same purpose or not

prem-k-r commented:
Checked in sandbox 
https://githubbox.com/Coderxrohan/materialYouNewTab/tree/main

Coderxrohan commented:
> Yes, it working, Great work! But there are lots of unnecessary codes

Hi Prem.
I made some changes now.
Removed that CSS or unwanted code from the main style.css file.
Reduced the unwanted code and made some changes in dark mode. js file

Have a look at it and suggest any changes, please!

prem-k-r commented:
Ok
I'll check it in the evening, rn I am busy in work

itz-rj-here commented:
Just need to check if it works properly on the firefox browser....

Coderxrohan commented:
> Just need to check if it works properly on the firefox browser....

I have tested it on Firefox, working normally.

itz-rj-here commented:
> > Just need to check if it works properly on the firefox browser....
> 
> I have tested it on Firefox, working normally.

fine then...

prem-k-r commented:
Actually, #284 was before the dark mode toogle, it was to switch between light theme and black theme.
I think, for now we can go with this, later we have to store last light theme and last dark theme, coz same color's dark mode one might not like it.

We will be removing dark mode toggle from there to place near color palatte without text.

> Auto dark mode should be hidden in dark mode on Google. If the user wants dark mode, only then will the Dakt mode schedule option be visible. This will save space in the control panel of the extension.

not really, like if someone want to edit in between,
but maybe we can make the time selector collapsable

prem-k-r commented:
As for code, you have html and css in js that can be easily move to resp. files

and better to use function-based approach over class-based approach.

code by chatgpt (need reworking and adjustment):
in css

```
.dark-mode-schedule-settings {
    margin-top: 15px;
    padding: 10px 15px;
    background-color: var(--accentLightTint-blue);
    border-radius: 15px;
    display: none;
}

.time-input {
    display: flex;
    align-items: center;
    justify-content: space-between;
    gap: 10px;
    margin-top: 10px;
}

.time-input input[type="time"] {
    padding: 5px 10px;
    border: 1px solid var(--darkColor-blue);
    border-radius: 8px;
    background-color: var(--whitishColor-blue);
    color: var(--textColorDark-blue);
    font-size: 14px;
}

.time-input input[type="time"]:focus {
    outline: none;
    border-color: var(--darkerColor-blue);
}
```

in dark-mode.js

```
document.addEventListener('DOMContentLoaded', () => {
    const storageKey = 'darkModeScheduleConfig';
    const defaultConfig = {
        enabled: false,
        start: '20:00',
        end: '06:00',
    };

    // DOM Elements
    const toggle = document.getElementById('darkModeSchedule');
    const settings = document.getElementById('darkModeScheduleSettings');
    const startInput = document.getElementById('darkModeStart');
    const endInput = document.getElementById('darkModeEnd');
    const darkModeToggle = document.getElementById('enableDarkModeCheckbox');

    let config = { ...defaultConfig };
    let checkInterval;

    function initializeScheduler() {
        // Load and apply settings
        const savedConfig = JSON.parse(localStorage.getItem(storageKey)) || defaultConfig;
        config = { ...config, ...savedConfig };

        // Set up UI
        toggle.checked = config.enabled;
        settings.style.display = config.enabled ? 'block' : 'none';
        startInput.value = config.start;
        endInput.value = config.end;

        // Add event listeners
        toggle.addEventListener('change', () => handleUpdate('toggle'));
        startInput.addEventListener('input', () => handleUpdate('time'));
        endInput.addEventListener('input', () => handleUpdate('time'));

        // Start schedule if enabled
        if (config.enabled) {
            updateDarkMode();
            startChecking();
        }
    }

    function handleUpdate(type) {
        if (type === 'toggle') {
            config.enabled = toggle.checked;
            settings.style.display = config.enabled ? 'block' : 'none';

            if (config.enabled) {
                updateDarkMode();
                startChecking();
            } else {
                stopChecking();
                setDarkMode(false);
            }
        } else if (type === 'time') {
            config.start = startInput.value;
            config.end = endInput.value;
            updateDarkMode();
        }

        saveConfig();
    }

    function updateDarkMode() {
        if (!config.enabled) return;

        const now = new Date();
        const currentTime = toMinutes(`${now.getHours()}:${now.getMinutes()}`);
        const startTime = toMinutes(config.start);
        const endTime = toMinutes(config.end);

        const isDark = startTime <= endTime
            ? currentTime >= startTime && currentTime <= endTime
            : currentTime >= startTime || currentTime <= endTime;

        setDarkMode(isDark);
    }

    function saveConfig() {
        localStorage.setItem(storageKey, JSON.stringify(config));
    }

    function setDarkMode(isActive) {
        document.documentElement.classList.toggle('dark-theme', isActive);
    }

    function toMinutes(time) {
        const [hours, minutes] = time.split(':').map(Number);
        return hours * 60 + minutes;
    }

    function startChecking() {
        if (checkInterval) clearInterval(checkInterval);
        checkInterval = setInterval(updateDarkMode, 60000); // Check every minute
    }

    function stopChecking() {
        if (checkInterval) clearInterval(checkInterval);
    }

    initializeScheduler();
});
```

in script.js nothing will be added

Coderxrohan commented:
> As for code, you have html and css in js that can be easily move to resp. files
> 
> and better to use function-based approach over class-based approach.
> 
> code by chatgpt (need reworking and adjustment): in css
> 
> ```
> .dark-mode-schedule-settings {
>     margin-top: 15px;
>     padding: 10px 15px;
>     background-color: var(--accentLightTint-blue);
>     border-radius: 15px;
>     display: none;
> }
> 
> .time-input {
>     display: flex;
>     align-items: center;
>     justify-content: space-between;
>     gap: 10px;
>     margin-top: 10px;
> }
> 
> .time-input input[type="time"] {
>     padding: 5px 10px;
>     border: 1px solid var(--darkColor-blue);
>     border-radius: 8px;
>     background-color: var(--whitishColor-blue);
>     color: var(--textColorDark-blue);
>     font-size: 14px;
> }
> 
> .time-input input[type="time"]:focus {
>     outline: none;
>     border-color: var(--darkerColor-blue);
> }
> ```
> 
> in dark-mode.js
> 
> ```
> document.addEventListener('DOMContentLoaded', () => {
>     const storageKey = 'darkModeScheduleConfig';
>     const defaultConfig = {
>         enabled: false,
>         start: '20:00',
>         end: '06:00',
>     };
> 
>     // DOM Elements
>     const toggle = document.getElementById('darkModeSchedule');
>     const settings = document.getElementById('darkModeScheduleSettings');
>     const startInput = document.getElementById('darkModeStart');
>     const endInput = document.getElementById('darkModeEnd');
>     const darkModeToggle = document.getElementById('enableDarkModeCheckbox');
> 
>     let config = { ...defaultConfig };
>     let checkInterval;
> 
>     function initializeScheduler() {
>         // Load and apply settings
>         const savedConfig = JSON.parse(localStorage.getItem(storageKey)) || defaultConfig;
>         config = { ...config, ...savedConfig };
> 
>         // Set up UI
>         toggle.checked = config.enabled;
>         settings.style.display = config.enabled ? 'block' : 'none';
>         startInput.value = config.start;
>         endInput.value = config.end;
> 
>         // Add event listeners
>         toggle.addEventListener('change', () => handleUpdate('toggle'));
>         startInput.addEventListener('input', () => handleUpdate('time'));
>         endInput.addEventListener('input', () => handleUpdate('time'));
> 
>         // Start schedule if enabled
>         if (config.enabled) {
>             updateDarkMode();
>             startChecking();
>         }
>     }
> 
>     function handleUpdate(type) {
>         if (type === 'toggle') {
>             config.enabled = toggle.checked;
>             settings.style.display = config.enabled ? 'block' : 'none';
> 
>             if (config.enabled) {
>                 updateDarkMode();
>                 startChecking();
>             } else {
>                 stopChecking();
>                 setDarkMode(false);
>             }
>         } else if (type === 'time') {
>             config.start = startInput.value;
>             config.end = endInput.value;
>             updateDarkMode();
>         }
> 
>         saveConfig();
>     }
> 
>     function updateDarkMode() {
>         if (!config.enabled) return;
> 
>         const now = new Date();
>         const currentTime = toMinutes(`${now.getHours()}:${now.getMinutes()}`);
>         const startTime = toMinutes(config.start);
>         const endTime = toMinutes(config.end);
> 
>         const isDark = startTime <= endTime
>             ? currentTime >= startTime && currentTime <= endTime
>             : currentTime >= startTime || currentTime <= endTime;
> 
>         setDarkMode(isDark);
>     }
> 
>     function saveConfig() {
>         localStorage.setItem(storageKey, JSON.stringify(config));
>     }
> 
>     function setDarkMode(isActive) {
>         document.documentElement.classList.toggle('dark-theme', isActive);
>     }
> 
>     function toMinutes(time) {
>         const [hours, minutes] = time.split(':').map(Number);
>         return hours * 60 + minutes;
>     }
> 
>     function startChecking() {
>         if (checkInterval) clearInterval(checkInterval);
>         checkInterval = setInterval(updateDarkMode, 60000); // Check every minute
>     }
> 
>     function stopChecking() {
>         if (checkInterval) clearInterval(checkInterval);
>     }
> 
>     initializeScheduler();
> });
> ```
> 
> in script.js nothing will be added

Thanks for reviewing code. I will make changes myself and updates it.

itz-rj-here commented:
@prem-k-r the branch has conflicts. Can you do something?

itz-rj-here commented:
@prem-k-r wsup with this pr?

prem-k-r commented:
> @prem-k-r wsup with this pr?

It's a great idea, we can go ahead.
Need refactoring, too much duplicate and unnecessary codes.
